crates/ruff_python_formatter/src/verbatim.rs

```diff
@@ -0,0 +1,505 @@
+                .last()
+                .map_or(statement.end(), Ranged::end);
+
+            verbatim_text(TextRange::new(format_off_comment.end(), end)).fmt(f)?;
+
+            return Ok(statement.statement());
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+enum InSuppression {
+    No,
+    Yes,
+}
+
+#[derive(Debug)]
+enum SuppressionComments<'a> {
+    /// The first `fmt: off` comment.
+    SuppressionStarts {
+        /// The comments appearing before the `fmt: off` comment
+        formatted_comments: &'a [SourceComment],
+        format_off_comment: &'a SourceComment,
+    },
+
+    /// A `fmt: on` comment inside a suppressed range.
+    SuppressionEnds {
+        /// The comments before the `fmt: on` comment that should *not* be formatted.
+        suppressed_comments: &'a [SourceComment],
+        format_on_comment: &'a SourceComment,
+
+        /// The comments after the `fmt: on` comment (if any), that should be formatted.
+        formatted_comments: &'a [SourceComment],
+
+        /// Any following `fmt: off` comment if any.
+        /// * `None`: The suppression ends here (for good)
+        /// * `Some`: A `fmt: off`..`fmt: on` .. `fmt: off` sequence. The suppression continues after
+        ///     the `fmt: off` comment.
+        format_off_comment: Option<&'a SourceComment>,
+    },
+
+    /// Comments that all fall into the suppressed range.
+    Suppressed { comments: &'a [SourceComment] },
+
+    /// Comments that all fall into the formatted range.
+    Formatted {
+        #[allow(unused)]
+        comments: &'a [SourceComment],
+    },
+}
+
+impl<'a> SuppressionComments<'a> {
+    fn unwrap_suppression_starts(&self) -> (&'a [SourceComment], &'a SourceComment) {
+        match self {
+            SuppressionComments::SuppressionStarts {
+                formatted_comments,
+                format_off_comment,
+            } => (formatted_comments, *format_off_comment),
+            _ => {
+                panic!("Expected SuppressionStarts")
+            }
+        }
+    }
+}
+
+struct CommentRangeIter<'a> {
+    comments: &'a [SourceComment],
+    source: &'a str,
+    in_suppression: InSuppression,
+}
+
+impl<'a> CommentRangeIter<'a> {
+    fn in_suppression(comments: &'a [SourceComment], source: &'a str) -> Self {
+        Self {
+            comments,
+            in_suppression: InSuppression::Yes,
+            source,
+        }
+    }
+
+    fn outside_suppression(comments: &'a [SourceComment], source: &'a str) -> Self {
+        Self {
+            comments,
+            in_suppression: InSuppression::No,
+            source,
+        }
+    }
+
+    /// Returns a slice containing the remaining comments.
+    fn as_slice(&self) -> &'a [SourceComment] {
+        self.comments
+    }
+}
+
+impl<'a> Iterator for CommentRangeIter<'a> {
+    type Item = SuppressionComments<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.comments.is_empty() {
+            None
+        } else {
```

Comment:
nit: dedent the body

```suggestion
        if self.comments.is_empty() {
            return None;
        }
```